public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx

```diff
@@ -319,6 +422,15 @@ export const QueryAndExpressionsStep = ({ editingExistingRule, onDataChange }: P
               Add query
             </Button>
           </Tooltip>
+          <SmartAlertTypeDetector
+            editingExistingRule={editingExistingRule}
+            rulesSourcesWithRuler={rulesSourcesWithRuler}
+            addExpressionsInQueries={addExpressionsInQueries}
```

Comment:
I feel like a lot of these functions we're passing down in to the `SmartAlertDetector` don't make sense, modifications to local state should happen where the state is managed and isn't a concern for this component. 

Instead we can pass a `onChangeType(type: RuleFormType) => void` handler in to the `SmartAlertTypeDetector` and do the modifications to the local state and reducer state in the `QueryAndExpressionsStep` component (since that one manages the form, reducer and local state).

Reply:
💯💯